tests/test_core_schedule.py

```diff
@@ -872,6 +872,22 @@ def test_multiple_services_are_present_in_schedule_after_adding(schedule, servic
     assert set(schedule.route_ids()) == expected_routes_after_adding
 
 
+def test_adding_services_from_schedule_does_not_change_route_data(
+        schedule, services_to_add):
+    routes_data_before_add = [r.__dict__ for r in services_to_add[0].routes()]
+    schedule_to_add = Schedule(services=services_to_add, epsg='epsg:27700')
+
+    schedule.add_services(list(schedule_to_add.services()))
+
+    for item in ['ordered_stops', 'arrival_offsets', 'departure_offsets']:
```

Comment:
Can we simplify this somehow? Ideally, I'd want to be able to do something like:

```python
routes_before_add = schedule.routes

schedule.add_services(services_to_add) # or schedule.add_services(schedule_to_add)

assert schedule.routes == routes_before_add 

```

If we don't want to change the API, maybe we can use test helper methods so we still get something like:

```python
assert_route_data_equal(schedule.routes, routes_before_add)
```

Reply:
I can simplify it to:
```
    schedule_to_add = Schedule(services=services_to_add, epsg='epsg:27700')
    route_ids_to_add = list(schedule_to_add.route_ids())
    routes_data_before_add = [schedule_to_add.route(r) for r in route_ids_to_add]

    schedule.add_services(list(schedule_to_add.services()))

    assert [schedule.route(r) for r in route_ids_to_add] == routes_data_before_add
```
but that does not compare the time offsets and trips. The things that are compared are only the list items inside the routes, so it would not be a generic `assert_route_data_equal` that could be reused, it would be `assert_route_list_data_equal` and used only once in this test.

Reply:
I know what the test is doing in general terms, but I find the detail of the assertions confusing. There is implicit knowledge of the fixtures being used in the assertions, e.g. `'new_route_0_0'` - nothing in the test reveals why we expect what we expect in the assertions.

The test should be more straightforward: grab the routes before, call the code under test, and compare the routes now to the routes before. That shouldn't require us to know anything about the actual _content_ of the routes. We don't care about the specifics; we only care that they haven't changed. We're doing work in the test that doesn't belong there.

The `Route` class has an equality comparator that you've alluded to. If it's missing some of the things that make a comparison between routes useful (time offsets and whatnot), should we add those? That feels like the place where the mechanics of route equality comparison should live.